extensions/ganglia-emitter/src/main/java/io/druid/emitter/ganglia/GangliaEmitter.java

```diff
@@ -0,0 +1,108 @@
+/*
+ *  Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ *  or more contributor license agreements. See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership. Metamarkets licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License. You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.druid.emitter.ganglia;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSortedSet;
+import com.metamx.common.logger.Logger;
+import com.metamx.emitter.core.Emitter;
+import com.metamx.emitter.core.Event;
+import com.metamx.emitter.service.ServiceMetricEvent;
+import info.ganglia.gmetric4j.gmetric.GMetric;
+import info.ganglia.gmetric4j.gmetric.GangliaException;
+
+import javax.validation.constraints.NotNull;
+import java.io.IOException;
+import java.net.InetAddress;
+
+/**
+ */
+public class GangliaEmitter implements Emitter
+{
+
+  private static Logger log = new Logger(GangliaEmitter.class);
+
+  private final GangliaEmitterConfig config;
+  private GMetric ganglia;
+
+  public GangliaEmitter(@NotNull GangliaEmitterConfig config)
+  {
+    this.config = config;
+  }
+
+  @Override
+  public void start()
+  {
+    try {
+      String spoof = InetAddress.getLocalHost().getHostAddress() + ":" + InetAddress.getLocalHost().getCanonicalHostName().split("\\.")[0];
+      log.info("Starting Ganglia Emitter. Spoof=[%s]", spoof);
+      this.ganglia = new GMetric(
+          config.getHostname(),
+          config.getPort(),
+          GMetric.UDPAddressingMode.UNICAST,
+          0,
+          true,
+          null,
+          spoof
+      );
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    log.info("constructed GMetric. host=[%s] port=[%s]", config.getHostname(), config.getPort());
+  }
+
+  @Override
+  public void emit(Event event)
+  {
+    if (event instanceof ServiceMetricEvent) {
+      ServiceMetricEvent metricEvent = (ServiceMetricEvent) event;
+
+      try {
+        ganglia.announce(
```

Comment:
@michaelschiff `emit` is called with the assumption that it is non-blocking and will not throw exceptions (assuming it's in the right state, like, it's okay if it throws an exception when it hasn't been started yet).

So usually the right way to implement it is to put events into a queue that is emptied by some other thread. If the queue is full then it's usually better to drop messages than to block (since blocking will block random threads in the rest of the process and grind the entire thing to a halt).